jax/numpy/lax_numpy.py

```diff
@@ -1768,16 +1760,23 @@ def identity(n, dtype=None):
 @_wraps(onp.arange)
 def arange(start, stop=None, step=None, dtype=None):
   lax._check_user_dtype_supported(dtype, "arange")
-  # If called like np.arange(N), we create a lazy lax._IotaConstant.
   if stop is None and step is None:
     dtype = dtype or _dtype(start)
-    if issubdtype(dtype, onp.integer):
-      return lax.iota(dtype, start)  # avoids materializing
-
-  # Fall back to instantiating an ndarray in host memory
-  dtype = dtype or result_type(
-    *(x for x in (start, stop, step) if x is not None))
-  return onp.arange(start, stop=stop, step=step, dtype=dtype)
+    return lax.iota(dtype, start)  # avoids materializing
+  elif step is None or onp.issubdtype(_dtype(step), onp.integer):
+    dtype = dtype or result_type(*(x for x in (start, stop, step)
+                                   if x is not None))
+    if stop is None:
+      start, stop = 0, int(start)
+    else:
+      start, stop = int(start), int(stop)
+    step = 1 if step is None else int(step)
+    quot, rem = divmod(stop - start, step)
+    num = quot + bool(rem)
+    return step * lax.iota(dtype, num) + start
```

Comment:
Just to confirm my understanding: this always materializes the `iota`, correct?

Reply:
Yes